kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KafkaRequestHandler.java

```diff
@@ -894,231 +893,84 @@ private void completeSendOperationForThrottling(long msgSize) {
-                return;
-            }
-            // add the topicPartition with timeout error if it's not existed in responseMap
-            produceRequest.partitionRecordsOrFail().keySet().forEach(topicPartition -> {
-                if (!responseMap.containsKey(topicPartition)) {
-                    responseMap.put(topicPartition, new PartitionResponse(Errors.REQUEST_TIMED_OUT));
+        String namespacePrefix = currentNamespacePrefix();
+        final AtomicInteger unfinishedAuthorizationCount = new AtomicInteger(numPartitions);
+        Consumer<Runnable> completeOne = (action) -> {
+            // When complete one authorization or failed, will do the action first.
+            action.run();
+            if (unfinishedAuthorizationCount.decrementAndGet() == 0) {
+                if (authorizedRequestInfo.isEmpty()) {
+                    resultFuture.complete(new ProduceResponse(unauthorizedTopicResponsesMap));
+                    return;
                 }
-            });
-            if (log.isDebugEnabled()) {
-                log.debug("[{}] Request {}: Complete handle produce.", ctx.channel(), produceHar.toString());
-            }
-            resultFuture.complete(new ProduceResponse(responseMap));
-        };
-        BiConsumer<TopicPartition, PartitionResponse> addPartitionResponse = (topicPartition, response) -> {
-            responseMap.put(topicPartition, response);
-            // reset topicPartitionNum
-            int restTopicPartitionNum = topicPartitionNum.decrementAndGet();
-            if (restTopicPartitionNum < 0) {
-                return;
-            }
-            if (restTopicPartitionNum == 0) {
-                complete.run();
+                AppendRecordsContext appendRecordsContext = AppendRecordsContext.get(
+                        topicManager,
+                        requestStats,
+                        this::startSendOperationForThrottling,
+                        this::completeSendOperationForThrottling,
+                        pendingTopicFuturesMap);
+                getReplicaManager().appendRecords(
+                        timeoutMs,
+                        false,
+                        produceHar.getRequest().version(),
+                        namespacePrefix,
+                        authorizedRequestInfo,
+                        appendRecordsContext
+                ).thenAccept(response -> {
+                    appendRecordsContext.recycle();
+                    Map<TopicPartition, PartitionResponse> mergedResponse = new HashMap<>();
+                    mergedResponse.putAll(response);
+                    mergedResponse.putAll(unauthorizedTopicResponsesMap);
+                    resultFuture.complete(new ProduceResponse(mergedResponse));
+                }).exceptionally(ex -> {
+                    appendRecordsContext.recycle();
+                    resultFuture.completeExceptionally(ex.getCause());
+                    return null;
```

Comment:
If you are sure that no exceptions could be thrown from `thenAccept` block, using `whenComplete` is better.

Otherwise, if exceptions were thrown from `thenAccept` block, `appendRecordsContext.recycle()` would be called twice.

Reply:
You're right. Addressed.